### PR TITLE
Snow 80090

### PR DESCRIPTION
**Pull request by smtakeda**
### Description
Send clientStartTime and retryCount with each /queries/v1/query-request

### Checklist
- [x] Code compiles correctly
- [x] Run ``make fmt`` to fix inconsistent formats
- [x] Run ``make lint`` to get lint errors and fix all of them
- [x] Created tests which fail without the change (if possible)
- [x] All tests passing
- [x] Extended the README / documentation, if necessary

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -25,60 +26,95 @@ func init() {
...
+	replace() *url.URL
 }
 
 // Make requestGUIDReplacer given a url string
-func makeRequestGUIDReplacer(url string) requestGUIDReplacerI {
-	startIndex := strings.Index(url, requestGUIDKey)
-	if startIndex == -1 {
-		return &transientReplacer{url}
+func makeRequestGUIDReplacer(urlPtr *url.URL) requestGUIDReplacerI {
+	_, err := url.ParseQuery(urlPtr.RawQuery)
```
Comment (ChTimTsubasa) [inline retry.go]: The reason to have a transientReplacer is to not even go through the url changing logic in the later retry if the url doesn't contains the requestGUID at the first place. So we might want to detect if the retryGuid key is in the query and if not just return the transientReplacer.

File: retry.go [hunk 2]
```diff
@@ -25,60 +26,95 @@ func init() {
...
-func (replacer *requestGUIDReplacer) replace() string {
-	return replacer.prefix + uuid.New().String() + replacer.suffix
+func (replacer *requestGUIDReplacer) replace() *url.URL {
+	vs, err := url.ParseQuery(replacer.urlPtr.RawQuery)
+	if err != nil {
+		return replacer.urlPtr
+	}
+	if len(vs.Get(requestGUIDKey)) == 0 {
+		return replacer.urlPtr
+	}
```
Comment (ChTimTsubasa) [inline retry.go]: The logic above should be moved to the makeRequestGUIDReplacer()

File: retry.go [hunk 3]
```diff
@@ -25,60 +26,95 @@ func init() {
...
+
+func (r *retryUpdate) replaceOrAdd(retry int) *url.URL {
+	if !strings.HasPrefix(r.TargetURL.Path, queryRequestPath) {
+		return r.TargetURL
+	}
+
+	vs, err := url.ParseQuery(r.TargetURL.RawQuery)
+	if err != nil {
+		return r.TargetURL
+	}
```
Comment (ChTimTsubasa) [inline retry.go]: I would probably either put these logic in the constructor and have a another struct named transienUpdate as consistent with GUIDReplacer and get slightly better performance in the case there are lot of retries, or just implement this as a simple function with no OOP at all.

File: retry.go [hunk 4]
```diff
@@ -25,60 +26,104 @@ func init() {
...
-	replace() string
+	replace() *url.URL
 }
 
 // Make requestGUIDReplacer given a url string
-func makeRequestGUIDReplacer(url string) requestGUIDReplacerI {
-	startIndex := strings.Index(url, requestGUIDKey)
-	if startIndex == -1 {
-		return &transientReplacer{url}
+func newRequestGUIDReplace(urlPtr *url.URL) requestGUIDReplacer {
```
Comment (smtakeda) [inline retry.go]: fixed the optimization issue.

File: retry.go [hunk 5]
```diff
@@ -25,60 +26,104 @@ func init() {
...
+
+// requestIDKey is attached to all requests to Snowflake
+const requestIDKey string = "requestId"
 
 // This class takes in an url during construction and replace the
 // value of request_guid every time the replace() is called
 // When the url does not contain request_guid, just return the original
 // url
-type requestGUIDReplacerI interface {
+type requestGUIDReplacer interface {
```
Comment (smtakeda) [inline retry.go]: changed the naming convention, i.e., "er"

File: retry.go [hunk 6]
```diff
@@ -25,60 +26,104 @@ func init() {
...
+	if !strings.HasPrefix(urlPtr.Path, queryRequestPath) {
+		// nop if not query-request
+		return &transientReplaceOrAdd{urlPtr}
+	}
+	values, err := url.ParseQuery(urlPtr.RawQuery)
+	if err != nil {
+		// nop if the URL is not valid
+		return &transientReplaceOrAdd{urlPtr}
+	}
+	return &retryCounterUpdate{urlPtr, values}
```
Comment (smtakeda) [inline retry.go]: fixed here too.

**Comment by ChTimTsubasa** [comment]
LGTM

**Pull request merged**